File: AOJ/GRL/1_A/1_A.rb
v, e, r = gets.chomp.split.map(&:to_i)
g = Array.new(v){ Array.new }
checked = Hash.new(0)
weight = Hash.new(0)
e.times do 
    s, t, d = gets.chomp.split.map(&:to_i)
    g[s] << t
    weight[[s, t]] = d
end

que = Array.new
que.push r
checked[r] = 0
while not que.empty? do
    node = que.shift    
    g[node].each do |new_node|
        next if not weight.has_key?([node, new_node])
        if checked.has_key?(new_node)
            # can find
            if checked[new_node] > checked[node] + weight[[node, new_node]]
                checked[new_node] = checked[node] + weight[[node, new_node]]
                que.push new_node
            end
        else
            # cannot find
            checked[new_node] = checked[node] + weight[[node, new_node]]
            que.push new_node
        end
    end
end

v.times do |i|
    if checked.has_key?(i)
        p checked[i]
    else 
        puts "INF"
    end
end

File: AOJ/GRL/1_A/1_A.cpp
#include<iostream>
#include<vector>
#include<map>
#include<queue>
#include<algorithm>
using namespace std;

int main(){
    int v, e, r;
    cin >> v >> e >> r;
    vector< vector<int> > g(v);
    map<pair<int, int>, int> weight;
    for(int i=0; i<e; i++){
        int s, t, d;
        cin >> s >> t >> d;
        g.at(s).push_back(t);
        g.at(t).push_back(s);
        weight[make_pair(s, t)] = d;
    }
    
    // BFS
    map<int, int> checked;
    queue<int> q;
    checked[r] = 0;
    q.push(r);
    while(!q.empty()){
        int node = q.front();
        q.pop();
        //for(auto new_node: g.at(node)){
        for(int i=0; i<g.at(node).size(); i++){
            int new_node = g.at(node).at(i);
            if(weight.find(make_pair(node, new_node)) == weight.end()) continue;
            if(checked.find(new_node) == checked.end()){
                // cannot find
                checked[new_node] = checked.at(node) + weight.at(make_pair(node, new_node));
                q.push(new_node);
            }
            else{
                // can find
                if(checked.at(new_node) > checked.at(node) + weight.at(make_pair(node, new_node))){
                    checked.at(new_node) = checked.at(node) + weight.at(make_pair(node, new_node));
                    q.push(new_node);
                }
            }
        }   
    }

    // DISP
    /*
    // disp graph
    for(int i=0; i<v; i++){
        cout << i << ':';
        for(auto node: g.at(i)){
            cout << node << ' ';
        }
        cout << endl;
    }
    */

    for(int i=0; i<v; i++){
        if(checked.find(i) == checked.end()) cout << "INF" << endl;
        else cout << checked.at(i) << endl;
    }
}